### PR TITLE
Kwargs to rasterio open

### LINKED ISSUE #3269: Accessing COG overviews with read_rasterio
[It's considered best practice to create cloud-optimized geotiff (COG) with overviews](https://medium.com/@_VincentS_/do-you-really-want-people-using-your-data-ec94cd94dc3f), which are essentially copies of the dataset at different resolutions to allow fast representation at different scales. 

It would be nice if we could pick a specific overview using `xr.read_rasterio`, perhaps by just an additional parameter on the call, like `xr.read_rasterio(url, overview=512)` or something.

Currently we need to use Rasterio to find out what the overlays are, for example:
```python
url = 'https://esip-pangeo-uswest2.s3-us-west-2.amazonaws.com/sciencebase/Southern_California_Topobathy_DEM_1m_cog.tif'
src = rasterio.open(url, 'r')
[src.overviews(i) for i in src.indexes]
``` 
which results in:
```
[[4, 8, 16, 32, 64, 128, 256, 512, 1023]]
```
See the notebook here for the hack workaround to extract an overlay from COG:
https://nbviewer.jupyter.org/gist/rsignell-usgs/dc4cf94fae97d085f6f2b9b896ec5336

### PR DESCRIPTION
**Pull request by pkopparla**
- [x] Closes #3269 
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

I've added a new `kwargs` argument to the function `xarray.open_rasterio` under `xarray/xarray/backends/rasterio_.py` which is passed on to the `rasterio.open` function as dicussed under issue #3269. I've added a line describing this change in `whats-new.rst`

I'm new to open source contributing, tried to follow instructions as much as I could. I am not sure how to write tests for this, would appreciate any help on that.

cheers,
Pushkar

### CODE REVIEW

File: doc/whats-new.rst [hunk 1]
```diff
@@ -53,6 +53,8 @@ New Features
 - Allow removal of the coordinate attribute ``coordinates`` on variables by setting ``.attrs['coordinates']= None``
   (:issue:`5510`).
   By `Elle Smith <https://github.com/ellesmith88>`_.
+- Added ``kwargs`` argument to xarray.open_rasterio to access overviews (`:issue:3269`).
```
Comment (dcherian) [inline doc/whats-new.rst:56]:
```suggestion
- Added ``kwargs`` argument to xarray.open_rasterio to access overviews (:issue:`3269`).
```

**Review by dcherian — COMMENTED** [review]
Thanks @pkopparla This looks very nice.

I too am not sure what a test would look like> Perhaps @scottyhq or @snowman2 have ideas,

**Comment by max-sixty** [comment]
Should this pass `**kwargs` through rather than take an explicit `kwargs` arg?

**Comment by pkopparla** [comment]
> Should this pass `**kwargs` through rather than take an explicit `kwargs` arg?

I am not sure about the nuances here, but the linked discussion shows a h5netcdf function that follows the same pattern.

**Comment by keewis** [comment]
I would try to stay as close to `open_dataset` as possible, which would make migrating to `rioxarray`'s engine easier once we deprecate `open_rasterio`. If I understand the signature of `open_dataset` correctly, this is called `backend_kwargs`?

**Comment by dcherian** [comment]
`rioxarray.open_rasterio` (and `xr.open_zarr`) uses the `**kwargs` format, so maybe that's the model to follow?

**Comment by keewis** [comment]
right, it seems it's both `backend_kwargs` and `**kwargs` for `open_dataset`.

File: doc/whats-new.rst [hunk 2]
```diff
@@ -53,6 +53,8 @@ New Features
 - Allow removal of the coordinate attribute ``coordinates`` on variables by setting ``.attrs['coordinates']= None``
   (:issue:`5510`).
   By `Elle Smith <https://github.com/ellesmith88>`_.
+- Added ``kwargs`` argument to xarray.open_rasterio to access overviews (:issue:`3269`).
+  By Pushkar Kopparla `<https://github.com/pkopparla>`_.
```
Comment (keewis) [inline doc/whats-new.rst:57]:
```suggestion
  By `Pushkar Kopparla <https://github.com/pkopparla>`_.
```

**Comment by snowman2** [comment]
For reference: 
- https://corteva.github.io/rioxarray/stable/examples/COG.html
- https://github.com/corteva/rioxarray/pull/48

**Comment by pkopparla** [comment]
Thanks everyone for the feedback. I've added a ``backend_kwargs`` and  a ``**kwargs`` argument following the `xarray.open_dataset` style.

**Comment by pkopparla** [comment]
Looks like the `backend_kwargs` argument is breaking the function. Should I just get rid of it and leave in `**kwargs`?

File: xarray/backends/rasterio_.py [hunk 3]
```diff
@@ -272,7 +280,14 @@ def open_rasterio(filename, parse_coordinates=None, chunks=None, cache=None, loc
...
         lock = RASTERIO_LOCK
 
-    manager = CachingFileManager(rasterio.open, filename, lock=lock, mode="r")
+    manager = CachingFileManager(
+        rasterio.open,
+        filename,
+        lock=lock,
+        mode="r",
+        backend_kwargs=backend_kwargs,
+        **kwargs,
```
Comment (keewis) [inline xarray/backends/rasterio_.py:289]: you would have to append `backend_kwargs` to `kwargs` (using `.update`) and then do
```suggestion
        kwargs=kwargs,
```

**Review by keewis — COMMENTED** [review]
> Should I just get rid of it and leave in `**kwargs`?

probably. @aurghs, @alexamici, it would be good to hear your thoughts on this if you have the time.

File: xarray/backends/rasterio_.py (hunk 3 above)
Comment (pkopparla) [reply xarray/backends/rasterio_.py:289]: Ok I'll just do this. Thanks @keewis

**Comment by pkopparla** [comment]
@dcherian @keewis Hey guys, a backend test is failing but I'm not sure why or how my change is affecting it. Would be grateful for a pointer.

**Comment by keewis** [comment]
this has been fixed / worked around on `main`, so the merge should have fixed it

**Comment by pkopparla** [comment]
@keewis thanks! Looks like most checks are successful now.

**Comment by keewis** [comment]
no, this is fine, and the PR looks good to me. I'd say let's wait one more day to give more time for reviews and then merge by tomorrow.

**Comment by aurghs** [comment]
> I would try to stay as close to `open_dataset` as possible, which would make migrating to `rioxarray`'s engine easier once we deprecate `open_rasterio`. If I understand the signature of `open_dataset` correctly, this is called `backend_kwargs`?

The idea was to deprecate `open_dataset signature` `backend_kwargs` in the future. Currently, in `open_dataset` signature, you can use either `backend_kwargs` or `kwargs` to bass the backend additional parameters. I would avoid adding `backend_kwargs` in `open_rasterio` interface.

File: doc/whats-new.rst [hunk 4]
```diff
@@ -97,6 +97,8 @@ New Features
 - Allow removal of the coordinate attribute ``coordinates`` on variables by setting ``.attrs['coordinates']= None``
   (:issue:`5510`).
   By `Elle Smith <https://github.com/ellesmith88>`_.
+- Added ``**kwargs`` argument to xarray.open_rasterio to access overviews (:issue:`3269`).
```
Comment (keewis) [inline doc/whats-new.rst:100]:
```suggestion
- Added ``**kwargs`` argument to :py:meth:`open_rasterio` to access overviews (:issue:`3269`).
```

**Review by keewis — COMMENTED** [review]
makes sense. @pkopparla, could you remove `backend_kwargs`?

**Comment by pkopparla** [comment]
@keewis Done!

**Comment by keewis** [comment]
thanks for the update, @pkopparla. I'll merge this tomorrow if there are no further comments.

**Comment by keewis** [comment]
Thanks, @pkopparla. I also noticed this is your first PR here. Welcome to `xarray`!

**Comment by pkopparla** [comment]
Thanks for leading me through the process @keewis , it's my first code contribution to any open-source project. Hope to do more in the future.

**Pull request merged**